Retain full path of files for directory uploads

bjoern-tantau opened the pull request:
To fix https://bugs.php.net/bug.php?id=77372 and improve support of `<input type="file" name="files" multiple webkitdirectory>` I introduced another item to the `$_FILES` array called `fullpath`, containing the full filename, as supplied by the user-agent.

nikic commented on main/rfc1867.c:
```diff
@@ -1149,9 +1149,39 @@ SAPI_API SAPI_POST_HANDLER_FUNC(rfc1867_post_handler) /* {{{ */
+			 */
+			if (!is_anonymous) {
+				safe_php_register_variable(lbuf, filename, strlen(filename), NULL, 0);
+			}
```
This is actually dead code. I've removed it in https://github.com/php/php-src/commit/61ba79df17867715b265de8b47238f96e62edf9c and you can drop it here as well (together with the snprintf above).

bjoern-tantau commented on main/rfc1867.c:
```diff
@@ -1149,9 +1149,39 @@ SAPI_API SAPI_POST_HANDLER_FUNC(rfc1867_post_handler) /* {{{ */
+				llen += MAX_SIZE_OF_INDEX + 1;
+			}
+
+			if (is_arr_upload) {
+				if (abuf) efree(abuf);
+				abuf = estrndup(param, strlen(param)-array_len);
+				snprintf(lbuf, llen, "%s_fullpath[%s]", abuf, array_index);
+			} else {
+				snprintf(lbuf, llen, "%s_fullpath", param);
+			}
```
@nikic Should I remove the whole block or just the else? In 61ba79df17867715b265de8b47238f96e62edf9c you left that in for `$foo_name`.

nikic replied:
You should drop the whole block, but you may need to move the abuf reallocation into the block below.

bjoern-tantau replied:
I removed the whole block altogether, including the abuf. It should already be handled by the `$foo_name` part above.

Thanks for your help, I'm more of a PHP and not a C developer.

nikic commented on main/rfc1867.c:
```diff
@@ -1138,20 +1135,42 @@ SAPI_API SAPI_POST_HANDLER_FUNC(rfc1867_post_handler) /* {{{ */
+				llen = (int)strlen(param);
+				lbuf = (char *) safe_erealloc(lbuf, llen, 1, MAX_SIZE_OF_INDEX + 1);
+				llen += MAX_SIZE_OF_INDEX + 1;
+			}
```
This code doesn't need to be repeated. You might have needed to adjust the value of MAX_SIZE_OF_INDEX, but it's currently `sizeof("[tmp_name]")`, which is the same as `sizeof("[fullpath]")`. Would need an adjustment if the name if changed though (to `[full_path]` for example).

nikic commented:
This looks reasonable to me. Main thing this needs is a rebase to drop the changes that are already on master, and a dedicated test for the new functionality. You've already updated plenty of tests that can serve as a template -- the `POST_RAW` section for a new test should be based on capturing what the browser actually sends in the `multiple webkitdirectory` case.

As a bit of bikeshed, should this be `full_path` rather than `fullpath` (in line with `tmp_name`)?

@cmb69 As you commented on the bug report, any opinion on this?

cmb69 commented:
Adding a new array member is fine for me. There is still the issue that IE provides the full path for single files, but that isn't a PHP issue when using a new member, and apparently [can be disabled](https://admx.help/?Category=Windows_10_2016&Policy=Microsoft.Policies.InternetExplorer::IZ_Policy_LocalPathForUpload_1) anyway. Besides that IE users likely have more serious issues.

bjoern-tantau commented:
> This looks reasonable to me. Main thing this needs is a rebase to drop the changes that are already on master, and a dedicated test for the new functionality. You've already updated plenty of tests that can serve as a template -- the `POST_RAW` section for a new test should be based on capturing what the browser actually sends in the `multiple webkitdirectory` case.
> 
> As a bit of bikeshed, should this be `full_path` rather than `fullpath` (in line with `tmp_name`)?

Done. I've renamed it to full_path (and updated MAX_SIZE_OF_INDEX), added a test and rebased on master.
